sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -1099,6 +1101,64 @@ object TypeCoercion {
         DateSub(l, Literal(days))
     }
   }
+
+  /**
+   * Coerces different children of Union to a common set of columns. Note that this must be
+   * run before `WidenSetOperationTypes`, because `WidenSetOperationTypes` should be run on
+   * correctly resolved column by name.
+   */
+  object UnionCoercion extends TypeCoercionRule {
+    private def unionTwoSides(
+        left: LogicalPlan, right: LogicalPlan, allowMissingCol: Boolean): LogicalPlan = {
+      val resolver = SQLConf.get.resolver
+      val leftOutputAttrs = left.output
+      val rightOutputAttrs = right.output
+
+      // Builds a project list for `right` based on `left` output names
+      val rightProjectList = leftOutputAttrs.map { lattr =>
+        rightOutputAttrs.find { rattr => resolver(lattr.name, rattr.name) }.getOrElse {
+          if (allowMissingCol) {
+            Alias(Literal(null, lattr.dataType), lattr.name)()
+          } else {
+            throw new AnalysisException(
+              s"""Cannot resolve column name "${lattr.name}" among """ +
+                s"""(${rightOutputAttrs.map(_.name).mkString(", ")})""")
+          }
+        }
+      }
+
+      // Delegates failure checks to `CheckAnalysis`
+      val notFoundAttrs = rightOutputAttrs.diff(rightProjectList)
+      val rightChild = Project(rightProjectList ++ notFoundAttrs, right)
+
+      // Builds a project for `logicalPlan` based on `right` output names, if allowing
+      // missing columns.
+      val leftChild = if (allowMissingCol) {
+        val missingAttrs = notFoundAttrs.map { attr =>
+          Alias(Literal(null, attr.dataType), attr.name)()
+        }
+        if (missingAttrs.nonEmpty) {
+          Project(leftOutputAttrs ++ missingAttrs, left)
+        } else {
+          left
+        }
+      } else {
+        left
+      }
+      Union(leftChild, rightChild)
+    }
+
+    override protected def coerceTypes(plan: LogicalPlan): LogicalPlan = plan resolveOperatorsUp {
+      case e if !e.childrenResolved => e
+
+      case Union(children, byName, allowMissingCol)
+          if byName =>
+        val union = children.reduceLeft { (left: LogicalPlan, right: LogicalPlan) =>
+          unionTwoSides(left, right, allowMissingCol)
```

Comment:
Just a question; the process of `unionTwoSides` is a type-coercion one? Seems like column resolution rather than type coercion.

Reply:
Seems not like typical type coercion case, but put here based on https://github.com/apache/spark/pull/28996#discussion_r453460284.

Reply:
hm I see.

Reply:
If looks not proper after rethinking, we can also move to other rule or create another rule.

Reply:
yea maybe a new analyzer rule.

Reply:
ok, moved.